src/theory/bags/bags_rewriter.cpp

```diff
@@ -0,0 +1,427 @@
+  switch (k)
+  {
+    case MK_BAG: return rewriteMakeBag(n);
+    case BAG_COUNT: return rewriteBagCount(n);
+    case UNION_MAX: return rewriteUnionMax(n);
+    case UNION_DISJOINT: return rewriteUnionDisjoint(n);
+    case INTERSECTION_MIN: return rewriteIntersectionMin(n);
+    case DIFFERENCE_SUBTRACT: return rewriteDifferenceSubtract(n);
+    case DIFFERENCE_REMOVE: return rewriteDifferenceRemove(n);
+    case BAG_CHOOSE: return rewriteChoose(n);
+    case BAG_CARD: return rewriteCard(n);
+    case BAG_IS_SINGLETON: return rewriteIsSingleton(n);
+    default: break;
+  }
+  return RewriteResponse(REWRITE_DONE, n);
+}
+
+RewriteResponse BagsRewriter::preRewrite(TNode n)
+{
+  Kind k = n.getKind();
+  switch (k)
+  {
+    case EQUAL: return rewriteEqual(n);
+    case BAG_IS_INCLUDED: return rewriteIsIncluded(n);
+    default: break;
+  }
+  return RewriteResponse(REWRITE_DONE, n);
+}
+
+RewriteResponse BagsRewriter::rewriteEqual(const TNode& n) const
+{
+  Assert(n.getKind() == EQUAL);
+  if (n[0] == n[1])
+  {
+    // (= A A) = true where A is a bag
+    NodeManager* nm = NodeManager::currentNM();
+    return RewriteResponse(REWRITE_DONE, nm->mkConst(true));
+  }
+  if (n[0].getKind() == MK_BAG && n[1].getKind() == MK_BAG)
+  {
+    //  (= (mkBag x c) (mkBag y d)) =
+    //  (ite
+    //        (and (<= c 0) (<= d 0))
+    //        true
+    //        (and (= x y) (= c d)))
+    Node x = n[0][0];
+    Node c = n[0][1];
+    Node y = n[1][0];
+    Node d = n[1][1];
+    NodeManager* nm = NodeManager::currentNM();
+    Node zero = nm->mkConst(Rational(0));
+    Node lte = nm->mkNode(LEQ, c, zero).andNode(nm->mkNode(LEQ, d, zero));
+    Node equal = x.eqNode(y).andNode(c.eqNode(d));
+    Node ite = nm->mkNode(ITE, lte, nm->mkConst(true), equal);
+    return RewriteResponse(REWRITE_AGAIN_FULL, ite);
+  }
+  return RewriteResponse(REWRITE_DONE, n);
+}
+
+RewriteResponse BagsRewriter::rewriteIsIncluded(const TNode& n) const
+{
+  Assert(n.getKind() == BAG_IS_INCLUDED);
+  // (bag.is_included A B) = ((difference_subtract A B) == emptybag)
+  NodeManager* nm = NodeManager::currentNM();
+  Node emptybag = nm->mkConst(EmptyBag(n[0].getType()));
+  Node subtract = nm->mkNode(DIFFERENCE_SUBTRACT, n[0], n[1]);
+  Node equal = subtract.eqNode(emptybag);
+  return RewriteResponse(REWRITE_AGAIN, equal);
+}
+
+RewriteResponse BagsRewriter::rewriteMakeBag(const TNode& n) const
+{
+  Assert(n.getKind() == MK_BAG);
+  // return emptybag for negative or zero multiplicity
+  if (n[1].isConst() && n[1].getConst<Rational>().sgn() != 1)
+  {
+    // (mkBag x c) = emptybag where c <= 0
+    NodeManager* nm = NodeManager::currentNM();
+    Node emptybag = nm->mkConst(EmptyBag(n.getType()));
+    return RewriteResponse(REWRITE_DONE, emptybag);
+  }
+  return RewriteResponse(REWRITE_DONE, n);
+}
+
+RewriteResponse BagsRewriter::rewriteBagCount(const TNode& n) const
+{
+  Assert(n.getKind() == BAG_COUNT);
+  if (n[1].isConst() && n[1].getKind() == EMPTYBAG)
```

Comment:
This should be generalized to handle all cases of constant arguments.  You should perhaps make a helper function `computeCount( Node x, Node b)` where x and b are constants.

Reply:
In this place I am assuming `x` is not a constant. I plan to implement `NormalForm::getNormalFrom` in another pull request to make this one easier.